sentry-graphql/src/main/java/io/sentry/graphql/ExceptionReporter.java

```diff
@@ -0,0 +1,146 @@
+import graphql.schema.DataFetchingEnvironment;
+import io.sentry.Hint;
+import io.sentry.IHub;
+import io.sentry.SentryEvent;
+import io.sentry.SentryLevel;
+import io.sentry.exception.ExceptionMechanismException;
+import io.sentry.protocol.Mechanism;
+import io.sentry.protocol.Request;
+import io.sentry.protocol.Response;
+import java.util.HashMap;
+import java.util.Map;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public final class ExceptionReporter {
+  private final boolean isSpring;
+
+  public ExceptionReporter(final boolean isSpring) {
+    this.isSpring = isSpring;
+  }
+
+  private static final @NotNull String MECHANISM_TYPE = "GraphqlInstrumentation";
+
+  public void captureThrowable(
+      final @NotNull Throwable throwable,
+      final @NotNull ExceptionDetails exceptionDetails,
+      final @Nullable ExecutionResult result) {
+    final @NotNull IHub hub = exceptionDetails.getHub();
+    final Mechanism mechanism = new Mechanism();
+    mechanism.setType(MECHANISM_TYPE);
+    mechanism.setHandled(false);
+    final Throwable mechanismException =
+        new ExceptionMechanismException(mechanism, throwable, Thread.currentThread());
+    final SentryEvent event = new SentryEvent(mechanismException);
+    event.setLevel(SentryLevel.FATAL);
+
+    final Hint hint = new Hint();
+    setRequestDetailsOnEvent(hub, exceptionDetails, event);
+
+    if (result != null) {
+      @NotNull Response response = new Response();
+      Map<String, Object> responseBody = result.toSpecification();
+      response.setData(responseBody);
+      event.getContexts().setResponse(response);
+    }
+
+    hub.captureEvent(event, hint);
+  }
+
+  private void setRequestDetailsOnEvent(
+      final @NotNull IHub hub,
+      final @NotNull ExceptionDetails exceptionDetails,
+      final @NotNull SentryEvent event) {
+    hub.configureScope(
+        (scope) -> {
+          final @Nullable Request scopeRequest = scope.getRequest();
+          if (scopeRequest != null) {
+            setDetailsOnRequest(hub, exceptionDetails, scopeRequest);
+            event.setRequest(scopeRequest);
+          } else {
+            Request newRequest = new Request();
+            setDetailsOnRequest(hub, exceptionDetails, newRequest);
+            event.setRequest(newRequest);
+          }
+        });
+  }
+
+  private void setDetailsOnRequest(
+      final @NotNull IHub hub,
+      final @NotNull ExceptionDetails exceptionDetails,
+      final @NotNull Request request) {
+    request.setApiTarget("graphql");
+
+    if (exceptionDetails.isSubscription() || !isSpring) {
+      final @NotNull Map<String, Object> data = new HashMap<>();
+
+      data.put("data", exceptionDetails.getQuery());
+
+      if (hub.getOptions().isSendDefaultPii()) {
+        data.put("variables", exceptionDetails.getVariables());
+      }
+
+      // for Spring this will be replaced by RequestBodyExtractingEventProcessor
+      request.setData(data);
```

Comment:
Why do we need to set it then?

Reply:
For non Spring and subscriptions

Reply:
is there a way to detect this in advance? I mean its ok being set just wanna be sure that this won't be "garbage" in some events that should not have a `data` at all, is that even possible?

Reply:
We are checking, see the if above `if (exceptionDetails.isSubscription() || !isSpring) {`

Reply:
If you know how, then you don't need to set `data` at all when you know its gonna be replaced, why setting it then?

Reply:
> If you know how, then you don't need to set data at all when you know its gonna be replaced, why setting it then?

I don't understand. We're only setting `data` if it's either a subscription or not running in spring-graphql. So it should never be replaced and the code should do exactly what you're asking for unless I misunderstand.

Reply:
@adinauer I am assuming because of the comment in the code:

> // for Spring this will be replaced by RequestBodyExtractingEventProcessor

Is the comment outdated then?

Reply:
I've updated the comment